semantic/src/main/kotlin/tools/samt/semantic/SemanticModelPostProcessor.kt

```diff
@@ -170,7 +170,69 @@ internal class SemanticModelPostProcessor(private val controller: DiagnosticCont
     }
 
     private fun checkRecord(record: RecordType) {
-        record.fields.forEach { checkModelType(it.type) }
+        record.fields.forEach {
+            checkModelType(it.type)
+            checkCycle(record, it)
+        }
+    }
+
+    private fun checkCycle(rootRecord: RecordType, rootField: RecordType.Field) {
+        fun impl(field: RecordType.Field, visited: List<Type>, encounteredOptional: Boolean = false) {
+            val typeReference = (field.type as? ResolvedTypeReference) ?: return
+            val type = typeReference.type
+            val record: RecordType
+            val newVisited: List<Type>
+            var isOptional = encounteredOptional || typeReference.isOptional
+            when (type) {
+                is RecordType -> {
+                    record = type
+                    newVisited = visited + record
+                }
+                is AliasType -> {
+                    val reference = type.fullyResolvedType ?: return
+                    val actualType = reference.type
+                    if (actualType !is RecordType) {
+                        return
+                    }
+                    record = actualType
+                    newVisited = visited + listOf(type, actualType)
+                    isOptional = isOptional || reference.isOptional
+                }
+                else -> return
+            }
+
+            if (record == rootRecord) {
+                val location = rootField.declaration.location
+                val path = newVisited.joinToString(" ► ") {
+                    buildString {
+                        append(it.humanReadableName)
+                        if (it is AliasType) {
+                            append(" (typealias)")
+                        }
+                    }
+                }
+                val context = controller.getOrCreateContext(location.source)
```

Comment:
You could use the new `.reportError` or `.reportWarning` extension function here